python-frontend/src/main/java/org/sonar/python/semantic/v2/types/FlowSensitiveTypeInference.java

```diff
@@ -59,40 +61,39 @@ public FlowSensitiveTypeInference(Set<SymbolV2> trackedVars,
   public ProgramState initialState() {
     TypeInferenceProgramState initialState = new TypeInferenceProgramState();
     for (SymbolV2 variable : trackedVars) {
-      PythonType pythonType = parameterTypesByName.get(variable.name());
-      initialState.setTypes(variable, pythonType != null ? Collections.singleton(pythonType) : Collections.emptySet());
+      var pythonTypeSet = Optional.of(variable.name())
+        .map(parameterTypesByName::get)
+        .map(Set::of)
+        .orElseGet(Set::of);
+      initialState.setTypes(variable, pythonTypeSet);
     }
     return initialState;
   }
 
   @Override
   public void updateProgramState(Tree element, ProgramState programState) {
     TypeInferenceProgramState state = (TypeInferenceProgramState) programState;
-    if (element.is(ASSIGNMENT_STMT)) {
-      AssignmentStatement assignment = (AssignmentStatement) element;
+    if (element instanceof AssignmentStatement assignment) {
       // update rhs
       updateTree(assignment.assignedValue(), state);
       handleAssignment(assignment, state);
       // update lhs
       assignment.lhsExpressions().forEach(lhs -> updateTree(lhs, state));
-    } else if (element.is(COMPOUND_ASSIGNMENT)) {
+    } else if (element instanceof CompoundAssignmentStatement) {
       // Assumption: compound assignments don't change types
       updateTree(element, state);
-    } else if (element.is(ANNOTATED_ASSIGNMENT)) {
```

Comment:
I wasn't actually sure about removing this bit, because I suspect it actually handles the case where we know the RHS type:
```
x: int = "hello" # We may not support type annotations in this iteration but we can still infer this is actually a string
```